From 17cd99e3811346417a5849960f8b5f358fe0358c Mon Sep 17 00:00:00 2001
From: marionm <mathis.marion@grenoble-inp.org>
Date: Tue, 24 Aug 2021 10:32:12 +0200
Subject: [PATCH 3/3] Kernel 4.4: use fwnode for irq domain

This patch follows a change to how IRQ are handled by HTVIC.
Here is a copy of the commit message for the corresponding HTVIC patch:

When creating an IRQ domain, the code used to store a device struct
pointer in the `of_node` field of the irq_domain struct returned by
`irq_domain_add_linear`. This was done by casting to a void pointer and
passing as the first argument of the function.

In the SPEC repository (and supposedly other depending repositories),
we would then get back the irq_domain struct by calling `irq_find_host`
with again a device struct pointer casted to a void pointer as the
argument. The function would compare the addresses of the 2 device
structs and return the right irq_domain.

This trick was most likely due to the fact that the IRQ domain API was
conceived around OpenFirmware before version 4.4, and that the project
should work for ACPI.
A workaround for kernel 4.7 was written, which involved using a `select`
function, and passing the address of the device struct as a parameter.
It was particularly ugly as it would require getting around the
`irq_find_host` to call immediately `irq_find_matching_fwspec` and pass
in the address of the wanted device struct as a parameter in a hacky way
to fit it inside two 32 bit integers.

Kernel version 4.4 introduced fwnodes, which would make easier working
with ACPI. Instead of repeating the hacky workaround (which would result
in a kernel error on later kernels when calling irq_domain_add_linear),
I allocated a fwnode_handle struct before creating a new IRQ domain.
Then I used the new irq_domain_create_* API to get an irq_domain using
this fwnode. I also took care of disallocating the fwnode_handle.

On the other end, we can just call `irq_find_matching_fwnode` and pass
dev.fwnode as a parameter.

Signed-off-by: Gwenhael GOAVEC <gwenhael.goavec@femto-st.fr>
Signed-off-by: Mathis MARION <mathis.marion@grenoble-inp.org>
---
 software/kernel/spec-core-fpga.c | 10 ++--------
 1 file changed, 2 insertions(+), 8 deletions(-)

diff --git a/software/kernel/spec-core-fpga.c b/software/kernel/spec-core-fpga.c
index b878f75..96f3db6 100644
--- a/software/kernel/spec-core-fpga.c
+++ b/software/kernel/spec-core-fpga.c
@@ -208,14 +208,8 @@ static struct resource spec_fpga_vic_res[] = {
 
 struct irq_domain *spec_fpga_irq_find_host(struct device *dev)
 {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,7,0)
-	struct irq_fwspec fwspec = {
-		.fwnode = dev->fwnode,
-		.param_count = 2,
-		.param[0] = ((unsigned long)dev >> 32) & 0xffffffff,
-		.param[1] = ((unsigned long)dev) & 0xffffffff,
-	};
-	return irq_find_matching_fwspec(&fwspec, DOMAIN_BUS_ANY);
+#if KERNEL_VERSION(4, 4, 0) <= LINUX_VERSION_CODE
+	return irq_find_matching_fwnode(dev->fwnode, DOMAIN_BUS_ANY);
 #else
 	return (irq_find_host((void *)dev));
 #endif
-- 
2.30.2

